# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/jsapi/PeerConnectionCtx.cpp
# Commit: 33ae2aabdf22
# Full Hash: 33ae2aabdf225a72dfceb09424e4e43ae9430e69
# Author: Atila Butkovits <abutkovits@mozilla.com>
# Date: 2025-06-12 16:31:09
# Regressor Bug: 1968812
# File Overlap Count: 1
# Description:
#   Revert "Bug 1968812 - Remove the CallWorker thread and use TaskQueue::Observer instead. r=bwc" for causing failures complaining about mutexLock.
#   
#   This reverts commit d53f26b530b4ad59aebe314c55a8af2c62a80fc3.
#   
#   Revert "Bug 1968812 - Copy WebrtcTaskQueueWrapper.h to WebrtcTaskQueueWrapper.cpp. r=bwc"
# ==============================================================================

diff -r ca232b0163c7 -r 33ae2aabdf22 dom/media/webrtc/jsapi/PeerConnectionCtx.cpp
--- a/dom/media/webrtc/jsapi/PeerConnectionCtx.cpp	Wed Jun 11 22:34:11 2025 +0300
+++ b/dom/media/webrtc/jsapi/PeerConnectionCtx.cpp	Wed Jun 11 23:24:24 2025 +0300
@@ -13,7 +13,7 @@
 #include "common/browser_logging/WebRtcLog.h"
 #include "gmp-video-decode.h"  // GMP_API_VIDEO_DECODER
 #include "gmp-video-encode.h"  // GMP_API_VIDEO_ENCODER
-#include "libwebrtcglue/WebrtcTaskQueueWrapper.h"
+#include "libwebrtcglue/CallWorkerThread.h"
 #include "modules/audio_device/include/fake_audio_device.h"
 #include "modules/audio_processing/include/audio_processing.h"
 #include "modules/audio_processing/include/aec_dump.h"
@@ -23,7 +23,6 @@
 #include "mozilla/glean/DomMediaWebrtcMetrics.h"
 #include "mozilla/dom/RTCStatsReportBinding.h"
 #include "nsCRTGlue.h"
-#include "nsIEventTarget.h"
 #include "nsIIOService.h"
 #include "nsIObserver.h"
 #include "nsIObserverService.h"
@@ -495,20 +494,27 @@
     audioStateConfig.audio_device_module =
         new rtc::RefCountedObject<FakeAudioDeviceModule>();
 
+    SharedThreadPoolWebRtcTaskQueueFactory taskQueueFactory;
     constexpr bool supportTailDispatch = true;
-    // This task queue is passed into libwebrtc by means of
-    // webrtc::TaskQueueBase::GetCurrent() while running on it.
+    // Note the NonBlocking DeletionPolicy!
+    // This task queue is passed into libwebrtc as a raw pointer.
     // WebrtcCallWrapper guarantees that it outlives its webrtc::Call instance.
-    // Outside of libwebrtc it works as a regular TaskQueue.
-    auto callWorkerThread = CreateWebrtcTaskQueueWrapper(
-        GetMediaThreadPool(MediaThreadType::WEBRTC_CALL_THREAD),
-        "CallWorker"_ns, supportTailDispatch);
+    // Outside of libwebrtc we must use ref-counting to either the
+    // WebrtcCallWrapper or to the CallWorkerThread to keep it alive.
+    auto callWorkerThread =
+        WrapUnique(taskQueueFactory
+                       .CreateTaskQueueWrapper<DeletionPolicy::NonBlocking>(
+                           "CallWorker", supportTailDispatch,
+                           webrtc::TaskQueueFactory::Priority::NORMAL,
+                           MediaThreadType::WEBRTC_CALL_THREAD)
+                       .release());
 
     UniquePtr<webrtc::FieldTrialsView> trials =
         WrapUnique(new MozTrialsConfig());
 
     mSharedWebrtcState = MakeAndAddRef<SharedWebrtcState>(
-        std::move(callWorkerThread), std::move(audioStateConfig),
+        new CallWorkerThread(std::move(callWorkerThread)),
+        std::move(audioStateConfig),
         already_AddRefed(CreateBuiltinAudioDecoderFactory().release()),
         std::move(trials));
     StartTelemetryTimer();